### PR TITLE
hotfix/DATA-3542

### PR DESCRIPTION
**Pull request by skavierin**
- Fixes to incremental build logic. Should now actually work as intended.
- docker.py cleanup, reduced branching and arg list for `.startup()` since almost all of the info is already within `target_adapter`

### CODE REVIEW

File: snowshu/core/docker.py [hunk 1]
```diff
@@ -144,20 +116,80 @@ def get_stopped_container(  # noqa pylint: disable=too-many-arguments
...
+                                         'already allocated, stop extra containers and rerun')
+                    raise
+
+                logger.info(
+                    f'Container {container.name} started, running initial setup...')
+                self._run_container_setup(container, target_adapter)
+                logger.info(f'Container {container.name} fully initialized.')
+
+                if len(arch_list) > 1:
+                    container.stop()
```
Comment (eculicny) [inline snowshu/core/docker.py:188]: Would it make sense to move this block of code back into something like the `get_stopped_container` function? It ends up being identical for both the `if` and the `else` blocks here.

File: snowshu/core/docker.py [hunk 2]
```diff
@@ -144,20 +116,80 @@ def get_stopped_container(  # noqa pylint: disable=too-many-arguments
...
+                logger.info(
+                    f"Creating stopped container {tagged_container_name}...")
+                self.remove_container(tagged_container_name)
+
+                container = self.client.containers.create(
+                    image.tags[0],
+                    target_adapter.DOCKER_START_COMMAND,
+                    network=network.name,
+                    name=tagged_container_name,
+                    hostname=target_adapter._credentials.host,  # noqa pylint: disable=protected-access
```
Comment (eculicny) [inline snowshu/core/docker.py:151]: Can we not access `target_adapter.credentials` here? I thought that was wrapped with property decorators/functions.

**Review by eculicny — COMMENTED** [review]
Looks good. I think we might be able to move some of the logic back into another function to reduce some identical code duplication.

File: snowshu/core/docker.py (hunk 2 above)
Comment (skavierin) [reply snowshu/core/docker.py:151]: We can, it's just that underscored one caught my eye first

File: snowshu/core/docker.py [hunk 3]
```diff
@@ -167,96 +199,75 @@ def get_stopped_container(  # noqa pylint: disable=too-many-arguments
...
+                    except AssertionError:
+                        logger.warning('Image tags do not match their actual architecture, '
+                                       'retag or delete postgres images manually to correct')
+
+                except ConnectionError as error:
+                    logger.error(
+                        'Looks like docker is not started, please start docker daemon\nError: %s', error)
+                    raise
+
+                tagged_container_name = f'snowshu_target_{arch}'
```
[the remark this answers is not in the document]
Comment (skavierin) [reply snowshu/core/docker.py:214]: We can't if we generalise this part as you asked below. This has to be target_adapter.credentials.hostname then

File: snowshu/core/docker.py (hunk 2 above)
Comment (eculicny) [reply snowshu/core/docker.py:151]: Makes sense. Let's use the actual `credentials` attribute in this case then.

File: snowshu/core/docker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (skavierin) [reply snowshu/core/docker.py]: Changed to `f'{hostname}_{arch}'`

File: snowshu/core/docker.py (hunk 1 above)
Comment (skavierin) [reply snowshu/core/docker.py]: Moved into other method, together with setup.

File: snowshu/adapters/target_adapters/base_target_adapter.py [hunk 4]
```diff
@@ -185,7 +187,7 @@ def initialize_replica(self,
                         f"The override image is not a Postgres image: {incremental_image}")
             except Exception as error:
```
Comment (ArtemHU) [inline snowshu/adapters/target_adapters/base_target_adapter.py:188]: should we define a custom exception for "image not available" case?
cc @eculicny

Comment (skavierin) [reply snowshu/adapters/target_adapters/base_target_adapter.py:188]: That's a year old code, I think better not touch it now

File: snowshu/adapters/target_adapters/base_target_adapter.py [hunk 5]
```diff
@@ -41,9 +42,10 @@ def __init__(self, replica_metadata: dict):
 
         self.target = DOCKER_TARGET_CONTAINER if IS_IN_DOCKER else 'localhost'
         self.credentials = self._generate_credentials(self.target)
+        self.shdocker = SnowShuDocker()
```
Comment (ArtemHU) [inline snowshu/adapters/target_adapters/base_target_adapter.py:45]: why don't to cast the SnowShuDocker class to static? cc @eculicny

**Review by eculicny — APPROVED** [review]
Looks good.

I think the updates suggested by @ArtemHU are probably worth looking into, but as separate pieces of work (out of scope of the hotfix)

**Pull request merged**